feat: opt-in sync of deletes and restores from web to Android

aleksandrsovtan opened the pull request:
## Description

Functionality for synchronizing the local media store when remotely deleting to trash and restoring files has been added. (Android)
This solves the issue of file deletion synchronization on web and mobile apps.

<!--- If it fixes an open issue, please link to the issue here. -->

Fixes # (issue)

## How Has This Been Tested?

To verify this, you need an Android 11+ device.  

1. Launch the Flutter app on an Android device.  
2. Log in to the app (I used the demo server and demo account).  
3. Go to **Settings** => **Advanced** and enable the **'Sync remote deletions'** switch.  
4. Download any photo or multiple photos from the photo list.  
5. Grant access to local albums, specifically to the **Immich** album (Cloud icon to the left of the profile icon).  
6. Open the **Immich** web app and log in with the same account (I used the demo).  
7. In the web app, delete (move to trash) a photo that is locally downloaded on your phone (marked with a cloud + icon).  
8. Open the gallery, go to the trash, and you should see the file that is in the trash on the web app.  
9. Return to the mobile app.  
10. In the web app, restore the file from the trash, and make sure it appears in the Flutter app's photo list (with the cloud + icon).  
11. Open the gallery, and the file should be in the **Immich**  folder (restored).

<!-- Images go below this line. -->

</details>

<!-- API endpoint changes (if relevant)
## API Changes
The `/api/something` endpoint is now `/api/something-else`
-->

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if applicable
- [ ] I have no unrelated changes in the PR.
- [ ] I have confirmed that any new dependencies are strictly necessary.
- [ ] I have written tests for new code (if applicable)
- [ ] I have followed naming conventions/patterns in the surrounding code
- [ ] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [ ] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

aleksandrsovtan committed:
Features: Local file movement to trash and restoration back to the album added. (Android)

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

alextran1502 commented on mobile/lib/widgets/settings/advanced_settings.dart:
```diff
@@ -45,6 +47,13 @@ class AdvancedSettings extends HookConsumerWidget {
+      SettingsSwitchListTile(
+        enabled: true,
+        valueNotifier: manageLocalMediaAndroid,
+        title: "Sync remote deletions",
```
Please put the string to en-us.json 

suggestion for the subtitle string

```
Automatically delete or restore an asset on this device when that action is taken on the web
```

alextran1502 replied:
Also, I would like to mark this as [EXPERIMENTAL]

alextran1502 commented:
Hello, I just ran the first round of functional testing on a Samsung S23, and I don't see the behavior I expected. Let me know if I misunderstood the PR's functionality.

1. Make sure that the option is enabled
2. Make sure the target asset is synced and present on the local device, meaning cloud + check icon.
3. Go to the web and trash the asset, confirm that the asset is in the trash
4. In the mobile app, I can see that the asset is in Immich's trash
5. I checked the native file explorer; in this case, it is the Samsung "My File" app. I checked the trash page, but I don't see any trash file there. 

Let me know if I have tested the PR correctly.

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan commented on mobile/lib/widgets/settings/advanced_settings.dart:
```diff
@@ -45,6 +47,13 @@ class AdvancedSettings extends HookConsumerWidget {
         title: "advanced_settings_troubleshooting_title".tr(),
         subtitle: "advanced_settings_troubleshooting_subtitle".tr(),
       ),
+      SettingsSwitchListTile(
```
You tested it a bit incorrectly. The Media Store API works with media files that are located in the gallery. So, after deletion, we need to check the trash in the gallery, not the trash in the file manager.  

When deleting a downloaded file in the web app, the file moves from the album to the trash (gallery), and when restoring it in the web app, it returns to the album from the trash.

aleksandrsovtan committed:
Comments fixes

aleksandrsovtan committed:
settings button marked as [EXPERIMENTAL]

shenlong-tanwen commented on mobile/lib/services/sync.service.dart:
```diff
@@ -748,9 +788,27 @@ class SyncService {
     return (existing, toUpsert);
   }
 
+  Future<void> _moveToTrashAssets(List<Asset> assetsList) async {
```
Why is the method named moveToTrash but instead doing both trashing and restoring the asset?

shenlong-tanwen commented on mobile/lib/services/sync.service.dart:
```diff
@@ -220,8 +231,31 @@ class SyncService {
   }
 
+  Future<void> _moveToTrashMatchedAssets(Iterable<String> idsToDelete) async {
+    final List<Asset> assets =
+        await _assetRepository.getAllByRemoteId(idsToDelete);
+
+    if (assets.isNotEmpty) {
+      final localAssets = await _assetRepository.getAllLocal();
+
+      // Find the local assets that match the remote assets by fileName
+      final matchedAssets = localAssets
+          .where(
+            (localAsset) => assets.any(
+              (remoteAsset) =>
+                  remoteAsset.fileName.contains(localAsset.fileName),
+            ),
+          )
+          .toList();
```
Why is this needed instead of fetching all local assets that also has a `remoteId` in the `idsToDelete` list?

aleksandrsovtan replied:
fixed

shenlong-tanwen reviewed (commented):
I'm not sure it is a good idea to try to keep the Trash inside Immich to be in sync with the one on the device. We can configure the number of days after which an asset is permanently deleted from the trash in Immich. However, the System trash on the device defaults to a 30 days window, after which the asset is removed permanently from the device. If the trash days is >30 inside Immich and the user tries to restore the asset after 30 days, there will be no asset, anymore on the device to be restored. It is not a problem with the _other app_ because it defaults to having 30 days as the trash window.

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
_moveToTrashMatchedAssets refactored, moveToTrash renamed.

benmccann commented:
> I'm not sure it is a good idea to try to keep the Trash inside Immich to be in sync with the one on the device. We can configure the number of days after which an asset is permanently deleted from the trash in Immich. However, the System trash on the device defaults to a 30 days window, after which the asset is removed permanently from the device.

This is a good catch. I do think it's kind of nice that it's viewable in the system trash though. I think there are a few options here:
- Immich does not use system trash and has its own trash
- Immich removes ability to configure trash length and uses 30 days like Android and iOS
- The photo is deleted on the mobile device at 30 days or Immich configured date - whichever is soonest. The image would probably be restorable from the server if it has a longer retention period 
- Same as the last option, but also cap the Immich retention at 30 days so the values do not conflict 

Do you have an opinion @alextran1502 ?

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

alextran1502 commented:
> Immich removes ability to configure trash length and uses 30 days like Android and iOS

I think this is the best compromise. The default is always to trash and have 30 days before deletion.

@jrasm91 any thoughts about this?

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

jrasm91 commented:
Personally, I think if you trash a file in Immich with this option enabled:
- The web should show a message similar to google photos
- The mobile app moves it to system trash
- The server deletes it after the configured trash days
- The mobile operating system deletes it following the system trash policy

Specifically when the Immich's trash policy in longer than 30 days I think it's fine for the item to be deleted on the device and only be restored on the web.

When the Immich trash policy is shorter than 30 days, I think it's also fine to leave the trashed photo in the mobile trash.

![image](https://github.com/user-attachments/assets/f9cfef4b-5657-4ecb-b47a-8329cfb9cb65)

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

benmccann commented:
I think that's probably pretty close to what happens with this PR today. I guess my question would be what happens if an asset is deleted from one trash and then restored from the other? Will it get synced to the other device? How and when does that happen? I think what Alex suggested is simpler conceptually in that it avoids cases like that.

It's been hard to keep this PR in sync with main and it keeps falling down the queue, so I sort of wonder if we shouldn't just merge it as is and figure out any follow ups separately.

alextran1502 committed:
Merge branch 'main' into mobile/manage_local_media_files

alextran1502 committed:
fix: bad merge

alextran1502 commented:
Hi @aleksandrsovtan, I just fixed the bad merge from `main` and reran some tests. I cannot trigger the condition of moving the deleted assets to either the gallery's trash or the file's trash.

Can you help me double-check? I am testing using Samsung S23.

Let get this done and have this PR merged

aleksandrsovtan commented:
> Hi @aleksandrsovtan, I just fixed the bad merge from `main` and reran some tests. I cannot trigger the condition of moving the deleted assets to either the gallery's trash or the file's trash.
> 
> Can you help me double-check? I am testing using Samsung S23.
> 
> Let get this done and have this PR merged


Here's the translation in English:

---

I tested on Android 14, Samsung Galaxy A15.

1. Build the app on the device.  
2. Download the images.  
3. Grant access to the album (in my case, it's Immich) by clicking the cloud icon near the profile in the top right corner.  
4. Ensure the images have synchronized (there should be a checkmark on the cloud icon).  
5. Log in to the same account on the web (I used the DEMO account).  
6. On the web version, delete an image that has been downloaded and synchronized locally.  
7. After moving it to the trash on the web, check the gallery in the app to see if the image has also moved to the trash (you may need to refresh the trash if the image doesn't appear, but the item count at the top should show an increase of one compared to before the deletion).  
8. Return from the gallery to the app on the Android device.  
9. Restore the image through the web — it should reappear in the Immich album.

Make sure that in the app settings, you have enabled the synchronization switch: **Settings => Advanced**.

alextran1502 commented:
@aleksandrsovtan The demo account cannot upload assets, so I am not sure how can you test it.

Do you mind record a screen capture video?

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan commented:
> @aleksandrsovtan The demo account cannot upload assets, so I am not sure how can you test it.
> 
> Do you mind record a screen capture video?

https://drive.google.com/drive/folders/1VT7MbljYLytJz4v2ROwNg2RUj7UImzOX?usp=drive_link

alextran1502 commented:
Thank you for the video. I still can't produce the expected functionality on my test instance.

The difference from my testing, which is the expected use case, is that I don't download the photos from the instance to my device, but I have the device upload existing photos to the instance. Then, I went on the web, deleted them, and ensured the feature was toggled on. However, none can be found in the Gallery app trash.

aleksandrsovtan committed:
Permission check and request for local storage added.

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

alextran1502 commented:
Hi @aleksandrsovtan, Thanks for the fix; I can now see the behavior in action! Great work

One change I would like to see is when the user toggles the feature; we request the file access permission right away. Then, only enable the feature if the permission is granted. This way, we can avoid weirdness happening in the app when the delete action is performed on the web and the app uses this mechanism for the first time.

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Permission request added on settings switcher

aleksandrsovtan commented:
> @aleksandrsovtan The demo account cannot upload assets, so I am not sure how can you test it.
> 
> Do you mind record a screen capture video?

Done

aleksandrsovtan commented:
> Hi @aleksandrsovtan, Thanks for the fix; I can now see the behavior in action! Great work
> 
> One change I would like to see is when the user toggles the feature; we request the file access permission right away. Then, only enable the feature if the permission is granted. This way, we can avoid weirdness happening in the app when the delete action is performed on the web and the app uses this mechanism for the first time.

Hi @alextran1502, done

alextran1502 commented:
@aleksandrsovtan I just tested the permission logic and saw the following behaviors

- Upon toggling the feature from the Immich app, the native permission prompt is presented as expected; however, when the user does not gain file access permission and returns to the app, the feature is still being turned on. I expect that if the user denies the permission, the feature will stay off.
- If the file access permission is gained, toggling the feature should not show the native permission prompt again. The native prompt is shown each time the feature switches from off -> on.

Can you help me resolve those two issues? We are very close now! 😃

shenlong-tanwen commented on mobile/android/app/src/main/kotlin/app/alextran/immich/MainActivity.kt:
```diff
@@ -3,13 +3,162 @@ package app.alextran.immich
         flutterEngine.plugins.add(BackgroundServicePlugin())
+
+        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL).setMethodCallHandler { call, result ->
+            when (call.method) {
+                "moveToTrash" -> {
+                    val fileName = call.argument<String>("fileName")
+                    if (fileName != null) {
```
Can you re-use the existing method channel in `BackgroundServicePlugin`?

shenlong-tanwen replied:
@aleksandrsovtan Can you resolve this comment to proceed with the review?

alextran1502 replied:
Hi @aleksandrsovtan, Can you help resolve this comment, moving the code to `BackgroundServicePlugin` so that we have better separation? I think after this resolved, we can emerge the PR. I just tested again and everything looks good

aleksandrsovtan replied:
This might be a challenge, since the background service plugin doesn't have an activity. I'll try to do it ASAP

aleksandrsovtan replied:
@shenlong-tanwen  Done!

shenlong-tanwen commented on mobile/android/app/src/main/kotlin/app/alextran/immich/MainActivity.kt:
```diff
@@ -3,13 +3,162 @@ package app.alextran.immich
+    }
+
+    fun getTrashedFileUri(contentResolver: ContentResolver, fileName: String): Uri? {
+        val contentUri = MediaStore.Files.getContentUri(MediaStore.VOLUME_EXTERNAL)
```
`MediaStore.VOLUME_EXTERNAL` required minSDK 29, however ours is set to 26. Have we tested this in Android 8 / 9?

aleksandrsovtan replied:
No, they haven't checked it, as I had warned and described in the testing method that for these methods to work, an Android 11+ device is required.

shenlong-tanwen replied:
> No, they haven't checked it, as I had warned and described in the testing method that for these methods to work, an Android 11+ device is required.

Yes, I think it has been decided to bump the minSdk to the minimum required version, i.e 30 and stop supporting the older versions of android. Can you handle this and other outstanding comments?

benmccann replied:
Sorry for the confusion, but instead of increasing the minimum version and dropping support for older versions of Android, let's just show the option for syncing deletes only when running Android 11

aleksandrsovtan replied:
I added a check that hides this switcher in the settings if the SDK version is below 30.

shenlong-tanwen commented on mobile/android/app/src/main/kotlin/app/alextran/immich/MainActivity.kt:
```diff
@@ -3,13 +3,162 @@ package app.alextran.immich
+        val queryArgs = Bundle().apply {
+            putString(ContentResolver.QUERY_ARG_SQL_SELECTION, "${MediaStore.Files.FileColumns.DISPLAY_NAME} = ?")
+            putStringArray(ContentResolver.QUERY_ARG_SQL_SELECTION_ARGS, arrayOf(fileName))
+            putInt(MediaStore.QUERY_ARG_MATCH_TRASHED, MediaStore.MATCH_ONLY)
```
`MediaStore.QUERY_ARG_MATCH_TRASHED` requires SDK 30

aleksandrsovtan replied:
same as on top comment

aleksandrsovtan replied:
I added a check that hides this switcher in the settings if the SDK version is below 30.

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Settings button logic changed

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan commented on mobile/android/app/src/main/kotlin/app/alextran/immich/MainActivity.kt:
```diff
@@ -3,13 +3,162 @@ package app.alextran.immich
 class MainActivity : FlutterActivity() {
+    private val CHANNEL = "file_trash"
+    private val REQUEST_MANAGE_STORAGE = 1001
+    private val REQUEST_TRASH_FILE = 1002
```
done

aleksandrsovtan commented:
> @aleksandrsovtan I just tested the permission logic and saw the following behaviors
> 
> * Upon toggling the feature from the Immich app, the native permission prompt is presented as expected; however, when the user does not gain file access permission and returns to the app, the feature is still being turned on. I expect that if the user denies the permission, the feature will stay off.
> * If the file access permission is gained, toggling the feature should not show the native permission prompt again. The native prompt is shown each time the feature switches from off -> on.
> 
> Can you help me resolve those two issues? We are very close now! 😃

@alextran1502 done!)

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

aleksandrsovtan committed:
Method channel file_trash moved to BackgroundServicePlugin

aleksandrsovtan committed:
Merge branch 'main' into mobile/manage_local_media_files

alextran1502 committed:
Merge branch 'main' into mobile/manage_local_media_files

BWagener commented:
I really hope there is a solution for the permission problem with the playstore.

Is there already an issue for this? Or is this the place to discuss it?

shenlong-tanwen commented:
> I really hope there is a solution for the permission problem with the playstore.

We are in the process of adding it back using a less intrusive permission on Android - #17828.

cchristchurch commented:
I have updated both the android app and the server to 1.132.3, however I still cannot see the experimental setting "Sync remote deletions" in the "Advanced" settings section. What am I missing? 

Note that my phone is still running Android 11, could that be the reason?

kujlek commented:
Hi, I've managed to reproduce the issue with latest immich server v2.6.3 and android app v2.6.3.

Web UI delete works ok (images are not re-uploaded) if these steps are taken:
- move images to trash
- sync immich app on phone
- empty the trash
- sync immich app on phone

Images are re-uploaded if immich app sync is not done between moving image to trash and emptying it.
